File: utils/benchmark/RC4/RC4.swift
@asmname("mach_absolute_time") func __mach_absolute_time__() -> UInt64

struct RC4 {
  var State : UInt8[]
  var I : UInt8 = 0
  var J : UInt8 = 0

  init() {
    State = new UInt8[256]
  }

  mutating
  func initialize(Key: UInt8[]) {
    for (var i = 0; i < 256; i++) {
      State[i] = UInt8(i)
    }

    var j : UInt8 = 0
    for (var i = 0; i < 256; i++) {
      var K : UInt8 = Key[i % Key.count]
      var S : UInt8 = State[i]
      j = j &+ S &+ K
      swapByIndex(i, y: Int(j))
    }
  }

  mutating
  func swapByIndex(x: Int, y: Int) {
    let T1 : UInt8 = State[x]
    let T2 : UInt8 = State[y]
    State[x] = T2
    State[y] = T1
  }

  mutating
  func next() -> UInt8 {
    I = I &+ 1
    J = J &+ State[Int(I)]
    swapByIndex(Int(I), y: Int(J))
    return State[Int(State[Int(I)] &+ State[Int(J)]) & 0xFF]
  }

  mutating
  func encrypt(inout Data: UInt8[]) {
    var cnt = Data.count
    for (var i = 0; i < cnt; i++) {
      Data[i] = Data[i] ^ next()
    }
  }
}

func benchRC4_internal(messageLen : Int, iterations : Int) {
  let Secret = "This is my secret message"
  let Key    = "This is my key"
  let SecretData : UInt8[] = Array(Secret.utf8)
  let KeyData    : UInt8[] = Array(Key.utf8)

  var LongData : UInt8[] = new UInt8[messageLen]

  // Generate a long message.
  for (var i = 0; i < messageLen; i++) {
    LongData[i] = SecretData[i % SecretData.count]
  }

  var Enc = RC4()
  Enc.initialize(KeyData)

  let start = __mach_absolute_time__()

  for (var i = 0; i < iterations; i++) {
    Enc.encrypt(&LongData)
  }

  let delta = __mach_absolute_time__() - start
  print("\(delta) nanoseconds. \(LongData[0])")
  print("\(Double(delta) / Double(iterations)) nanoseconds/lap")
}


func benchRC4() {
  benchRC4_internal(5000, 100000)
}

benchRC4()

File: utils/benchmark/RC4/RC4.py

class RC4:
  def __init__(self):
    self.state = [0] * 256
    self.I = 0
    self.J = 0

  def init(self, key):
    for i in xrange(256):
      self.state[i] = i

    j = 0
    for i in xrange(256):
      K = ord(key[i % len(key)])
      S = self.state[i]
      j = (j + S + K) % 256
      self.swapByIndex(i, j)

  def swapByIndex(self, i, j):
    self.state[i], self.state[j] = self.state[j], self.state[i]

  def next(self):
    self.I = (self.I + 1) % 256
    self.J = (self.J + self.state[self.I]) % 256
    self.swapByIndex(self.I, self.J)
    return self.state[(self.state[self.I] + self.state[self.J]) % 256]

  def encrypt(self, data):
    for i in xrange(len(data)):
      data[i] = data[i] ^ self.next()


def benchRC4_internal(messageLen, iterations):
  Secret = "This is my secret message"
  Key    = "This is my key"
  LongData = [0] * messageLen

  for i in xrange(messageLen):
    LongData[i] = ord(Secret[i % len(Secret)])

  Enc = RC4()
  Enc.init(Key)

  for i in xrange(iterations):
    Enc.encrypt(LongData)

benchRC4_internal(5000, 100000)
